mixer/pkg/checkcache/keyShape.go

```diff
@@ -0,0 +1,255 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checkcache
+
+import (
+	"encoding/binary"
+	"math"
+	"sort"
+	"time"
+
+	"crypto/md5"
+
+	mixerpb "istio.io/api/mixer/v1"
+	"istio.io/istio/mixer/pkg/attribute"
+	"istio.io/istio/mixer/pkg/pool"
+)
+
+// A reference to an attribute and to an optional map key.
+type attributeRef struct {
+	// name of the attribute
+	name string
+
+	// only used if attribute is a stringMap
+	mapKey string
+}
+
+// keyShape describes the attributes that must/must not participate in key formation
+type keyShape struct {
+	// These attributes must be absent from the input bag
+	absentAttrs []attributeRef
+
+	// These attributes must be present in the input bag
+	presentAttrs []attributeRef
+}
+
+const (
+	delimiter = 0
+)
+
+func newKeyShape(ra mixerpb.ReferencedAttributes, globalWords []string) keyShape {
+	ks := keyShape{}
+
+	for _, match := range ra.AttributeMatches {
+		var ar attributeRef
+
+		ar.name = getString(match.Name, globalWords, ra.Words)
+		ar.mapKey = getString(match.MapKey, globalWords, ra.Words)
+
+		if match.Condition == mixerpb.ABSENCE {
+			ks.absentAttrs = append(ks.absentAttrs, ar)
+		} else if match.Condition == mixerpb.EXACT {
+			ks.presentAttrs = append(ks.presentAttrs, ar)
+		}
+	}
+
+	sort.Slice(ks.absentAttrs, func(i int, j int) bool {
+		return ks.absentAttrs[i].name < ks.absentAttrs[j].name
+	})
+
+	sort.Slice(ks.presentAttrs, func(i int, j int) bool {
+		return ks.presentAttrs[i].name < ks.presentAttrs[j].name
+	})
+
+	return ks
+}
+
+// isCompatible determines whether the input bag meets the requirements to be used
+// with this instance
+func (ks keyShape) isCompatible(attrs attribute.Bag) bool {
+	return ks.checkAbsentAttrs(attrs) && ks.checkPresentAttrs(attrs)
+}
+
+// checkAbsentAttrs ensures that the input bag doesn't include any of the attributes that must be absent
+func (ks keyShape) checkAbsentAttrs(attrs attribute.Bag) bool {
+	for i := 0; i < len(ks.absentAttrs); i++ {
+		name := ks.absentAttrs[i].name
+
+		v, ok := attrs.Get(name)
+		if !ok {
+			// it's absent, all good
+			continue
+		}
+
+		sm, ok := v.(map[string]string)
+		if !ok {
+			// attribute is present and it's not a string map, so this bag isn't compatible
+			return false
+		}
+
+		// since absentAttrs is sorted by name, continue processing stringMaps until a new name is found.
+		for {
+			_, ok := sm[ks.absentAttrs[i].mapKey]
+			if ok {
+				// if the map key is present, then this bag won't work
+				return false
+			}
+
+			// break loop if at the end or name changes.
+			if i+1 == len(ks.absentAttrs) || ks.absentAttrs[i+1].name != name {
+				break
+			}
+			i++
+		}
+	}
+
+	// compatible
+	return true
+}
+
+// checkPresentAttrs ensures that the input bag includes all of the attributes that must be present
+func (ks keyShape) checkPresentAttrs(attrs attribute.Bag) bool {
```

Comment:
is there any diff between this method and `checkAbsentAttrs` other than the `[] attributeRefs` being used for comparison? Can we merge these into:

```
func (ks keyShape) checkAbsentAttrs(attrs attribute.Bag) bool {
   return checkAttrs(ks.absentAttrs, attrs)
}

func (ks keyShape) checkPresentAttrs(attrs attribute.Bag) bool {
   return checkAttrs(ks.presentAttrs, attrs)
}
```